msmbuilder/featurizer/featurizer.py

```diff
@@ -200,6 +200,61 @@ def partial_transform(self, traj):
         return x
 
 
+
+class StrucRMSDFeaturizer(Featurizer):
+    """Featurizer based on overall RMSD to a set of reference structures.
+
+    Parameters
+    ----------
+    atom_indices : np.ndarray, shape=(n_atoms,), dtype=int
+        The indices of the atoms to superpose and compute the distances with
+    reference_traj : md.Trajectory
+        The reference conformation to superpose each frame with respect to
+        (only the first frame in reference_traj is used)
+    superpose_atom_indices : np.ndarray, shape=(n_atoms,), dtype=int
+        If not None, these atom_indices are used for the superposition
+    """
+
+    def __init__(self, atom_indices, reference_traj, superpose_atom_indices=None):
+        self.atom_indices = atom_indices
+        if superpose_atom_indices is None:
+            self.superpose_atom_indices = atom_indices
+        else:
+            self.superpose_atom_indices = superpose_atom_indices
+        self.reference_traj = reference_traj
+#        self.n_features = len(self.atom_indices)
+
+    def partial_transform(self, traj):
+        """Featurize an MD trajectory into a vector space via distance
+        after superposition
+
+        Parameters
+        ----------
+        traj : mdtraj.Trajectory
+            A molecular dynamics trajectory to featurize.
+
+        Returns
+        -------
+        features : np.ndarray, dtype=float, shape=(n_samples, n_features)
+            A featurized trajectory is a 2D array of shape
+            `(length_of_trajectory x n_features)` where each `features[i]`
+            vector is computed by applying the featurization function
+            to the `i`th snapshot of the input trajectory.
+
+        See Also
+        --------
+        transform : simultaneously featurize a collection of MD trajectories
+        """
+        traj.superpose(self.reference_traj, atom_indices=self.superpose_atom_indices)
+        x = []
+        for i in range(0, self.reference_traj.n_frames):
+            y = md.rmsd(traj, self.reference_traj, i)
+            x.append(y)
```

Comment:
This `for` loop will be needlessly slow, because rmsd has been highly optimized but there's a lot of overhead for python looping. This is why numpy operations that act on entire vectors at once are much faster than pure python loops.


It's probably good enough for now, but we could do something faster. This calculation is analgous to a function called `cdist` in the scipy library ([docs](http://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.spatial.distance.cdist.html)). We have RMSD and Trajectory-aware analogs of many of the functions in that module in the subpackage `msmbuilder.libdistance` -- e.g. `pdist` [here](https://github.com/msmbuilder/msmbuilder/blob/master/msmbuilder/libdistance/libdistance.pyx#L132-L157), but it looks like we don't have one for cdist. 

------

For example, with this code on my laptop, 

```
# test_timing.ipy
import numpy as np

def add_arrays_python(a, b):
    assert len(a) == len(b)
    result = np.empty_like(a)
    size = len(a)

    for i in range(size):
        result[i] = a[i] + b[i]

    return result


def add_arrays_numpy(a, b):
    assert len(a) == len(b)

    result = a + b
    
a = np.random.randn(10000)
b = np.random.randn(10000)
%timeit add_arrays_python(a, b)
%timeit add_arrays_numpy(a, b)
```

```
$ ipython  test_timing.ipy
100 loops, best of 3: 3.57 ms per loop
The slowest run took 5.73 times longer than the fastest. This could mean that an intermediate result is being cached
100000 loops, best of 3: 6.84 µs per loop
```
So the numpy code is ~500x faster at computing the same result.